File: GK_Updater_Module/ProgressCounter.vb
﻿Public Class ProgressCounter
    Private _progBytesMoved As Integer
    Private _progTotalBytes As Integer
    Private _speedBytesMoved As Integer
    Private _currentTick As Integer
    Private _startTick As Integer
    Private _speedThroughput As Double
    Sub New()
        _progBytesMoved = 0
        _progTotalBytes = 0
        _speedBytesMoved = 0
        _currentTick = 0
        _startTick = 0
        _speedThroughput = 0
    End Sub
    Public Property BytesToTransfer As Integer
        Get
            Return _progTotalBytes
        End Get
        Set(value As Integer)
            _progTotalBytes = value
        End Set
    End Property
    Public ReadOnly Property Percent As Integer
        Get
            Return CInt((_progBytesMoved / _progTotalBytes) * 100)
        End Get
    End Property
    Public Property BytesMoved As Integer
        Get
            Return _progBytesMoved
        End Get
        Set(value As Integer)
            _speedBytesMoved += value
            _progBytesMoved += value
        End Set
    End Property
    Public ReadOnly Property Throughput As Double
        Get
            Return _speedThroughput
        End Get
    End Property
    Public Sub ResetProgress()
        _progBytesMoved = 0
    End Sub
    Public Sub Tick()
        _currentTick = Environment.TickCount
        If _startTick > 0 Then
            If _speedBytesMoved > 0 Then
                Dim elapTime = _currentTick - _startTick
                _speedThroughput = Math.Round((_speedBytesMoved / elapTime) / 1000, 2)
            End If
        Else
            _startTick = _currentTick
        End If
    End Sub
End Class

File: GKUpdaterLibC/ProgressCounter.cs
using System;

namespace RemoteFileTransferTool
{
    public class ProgressCounter
    {
        #region "Fields"

        private long _currentTick;
        private long _progBytesMoved;
        private long _progTotalBytes;
        private long _speedBytesMoved;
        private double _speedThroughput;

        private long _startTick;

        #endregion "Fields"

        #region "Constructors"

        public ProgressCounter()
        {
            _progBytesMoved = 0;
            _progTotalBytes = 0;
            _speedBytesMoved = 0;
            _currentTick = 0;
            _startTick = 0;
            _speedThroughput = 0;
        }

        #endregion "Constructors"

        #region "Properties"

        public long BytesMoved
        {
            get { return _progBytesMoved; }
            set
            {
                _speedBytesMoved += value;
                _progBytesMoved += value;
            }
        }

        public long BytesToTransfer
        {
            get { return _progTotalBytes; }
            set { _progTotalBytes = value; }
        }

        public int Percent
        {
            get
            {
                if (_progTotalBytes > 0)
                {
                    return (int)Math.Round((((float)_progBytesMoved / (float)_progTotalBytes) * 100), 0);
                }
                else
                {
                    return 0;
                }
            }
        }

        public double Throughput
        {
            get { return _speedThroughput; }
        }

        #endregion "Properties"

        #region "Methods"

        public void ResetProgress()
        {
            _progBytesMoved = 0;
        }

        public void Tick()
        {
            _currentTick = DateTime.Now.Ticks;
            if (_startTick > 0)
            {
                if (_speedBytesMoved > 0)
                {
                    double elapTime = _currentTick - _startTick;
                    _speedThroughput = Math.Round((_speedBytesMoved / (elapTime / 10000)) / 1000, 2);
                }
            }
            else
            {
                _startTick = _currentTick;
            }
        }

        #endregion "Methods"
    }
}